src/storage/txn/actions/prewrite.rs

```diff
@@ -1,135 +1,490 @@
 // Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
-use crate::storage::mvcc::{
-    metrics::{MVCC_CONFLICT_COUNTER, MVCC_DUPLICATE_CMD_COUNTER_VEC},
-    ErrorInner, LockType, MvccTxn, Result as MvccResult,
+use crate::storage::{
+    mvcc::{
+        metrics::{
+            CONCURRENCY_MANAGER_LOCK_DURATION_HISTOGRAM, MVCC_CONFLICT_COUNTER,
+            MVCC_DUPLICATE_CMD_COUNTER_VEC,
+        },
+        ErrorInner, Lock, LockType, MvccTxn, Result,
+    },
+    txn::actions::check_data_constraint::check_data_constraint,
+    txn::LockInfo,
+    Snapshot,
 };
-use crate::storage::txn::actions::check_data_constraint::check_data_constraint;
-use crate::storage::txn::actions::shared::prewrite_key_value;
-use crate::storage::Snapshot;
-use txn_types::{Key, Mutation, TimeStamp, WriteType};
+use fail::fail_point;
+use std::cmp;
+use txn_types::{is_short_value, Key, Mutation, MutationType, TimeStamp, Value, Write, WriteType};
 
+/// Prewrite a single mutation by creating and storing a lock and value.
 pub fn prewrite<S: Snapshot>(
```

Comment:
I prefer keep `txn` as the first argument just as other actions does, do you have other considerations? @nrc

Reply:
seems fine, consistency is good!